### PR TITLE
Jc jupyter extension

### PR DESCRIPTION
**Pull request by jdcanas**
Initial commit for the jupyter UI extension to get some eyes on it.

Joel, I modified the extension you made so its toggleable based on the state stored in welder as opposed to an on-load event. This is relevant because users can go to playground mode after a time period of being in edit mode. Still need to actually call welder from it though. 

I still need to:
   -get product sign-off on UX
   -remove the console.logs
  ~~-verify the proper URLs to use for jupyter server and welder (im hoping localhost will work)~~
   -test in jupyter image
   -test in fiab with jupyter image and welder image running
      -payloads may change 

In all cases:

- [x] Get a thumbsworth of review and PO signoff if necessary
- [x] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/jupyter/safe-mode.js [hunk 1]
```diff
@@ -14,88 +14,43 @@
...
+        return true //TODO figure out what the welder link will be relative to the jupyter image
+    }
 
-    $('#notification_area').prepend(
-        '<div id="safe-mode" style="background-color: #FFFFB2;" ' +
-          'class="notification_widget btn btn-xs navbar-btn">' +
-          '<span>Safe Mode - your edits will not be saved.' +
-            '</span></div>');
-  };
+    function toggleUIControls(shouldHide) {
```
Comment (jmthibault79) [inline src/main/resources/jupyter/safe-mode.js]: Oh this is much nicer.  Thanks.

File: src/main/resources/jupyter/safe-mode.js [hunk 2]
```diff
@@ -14,88 +14,43 @@
...
+        })
+    }
 
-    // TODO: convert to tooltip with this text:
-    // Safe Mode allows you to explore, change, and run the code,
-    // but your edits will not be saved.
-    // To save your work, choose Make a Copy from the File menu to
-    // make your own version.
+    function isSafeMode() {
+        return true //TODO figure out what the welder link will be relative to the jupyter image
```
Comment (jmthibault79) [inline src/main/resources/jupyter/safe-mode.js]: Let's make this false until it's working correctly

File: src/main/resources/jupyter/safe-mode.js [hunk 3]
```diff
@@ -14,88 +14,43 @@
...
+    function checkMetaLoop() {
+        toggleUIControls(isSafeMode())
 
-    $('#notification_area')
-        .append(
-            '<style>' +
-              // e.g. Notebook Saved
-              '#notification_notebook span { display: none; } ' +
-              '</style>');
+        const interval = setInterval(() => {
```
Comment (jmthibault79) [inline src/main/resources/jupyter/safe-mode.js]: Is this going to change during the life of a notebook?  If not, we don't need to poll.

When we transition between safe mode and other modes, we will be copying the notebook file to a new directory.  Will that re-initialize this plugin, or will the current instance survive that transition?

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        },
+        'Reload the workspace version and discard your changes': {
+            click: () => updateLocalCopyWithRemote()
+        }
+    }
+
+    const lockIssueButtons = {
+        'Run in Playground Mode': {
+            click: () => openPlaygroundMode(),
+            'class': 'btn-primary'
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: Why does this need a class but not the others?

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    const modeBannerId = "notification_mode"
+    const lockConflictTitle = "File is in use"
+    const syncIssueTitle = "File versions out of sync"
+    const syncIssueBody = "Your version of this file does not match the version in the workspace. What would you like to do?"
+    const syncIssueFatalBody = "This file was either deleted or never was stored with us."
+
+    //TODO URL resolution
+    const leoUrl = 'http://localhost:8080'
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: Because this is client-side, localhost won't work.

The browser running the notebook will set an auth cookie so we will be able to use relative paths here, which will call out to the Leo proxy.  `welderUrl` is `proxy/<billing project>/<cluster name>/welder`

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        'Access-Control-Allow-Headers': '*'
+    }
+
+    const basePayload = {
+        // mode: "no-cors",
+        headers: headers
+    }
+
+    //TODO resolve these links
+    const jupyterBaseUrl = "http://localhost:8000/"
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: Like above, we need to use relative path to Leo.  This is simply `notebooks`.

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    const jupyterContentsAPIUrl = jupyterBaseUrl + "api/contents/"
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: What is this?

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+    async function checkMeta() {
+        const payload = Object.assign(basePayload, {
+            body: JSON.stringify(getLocalPath()),
+            method: 'POST'
+        })
+
+        return fetch(checkMetaUrl, payload)
+            .then(res => {
+                console.log('in first then of checkmeta', res)
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: rm console.logs

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    //this function assumes any status not included in these lists represents an in sync notebook to defend against future fields being added being auto-categorized as failures
+    function handleCheckMetaResp(res) {
+        meta = res //set meta state
+
+        const outOfSyncStatuses = ["DESYNCHRONIZED", "LOCAL_CHANGED", "REMOTE_CHANGED"]
+        const fatalSyncStatuses = ["REMOTE_NOT_FOUND"]
+
+        if (outOfSyncStatuses.includes(res.syncStatus)) {
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: which of these 2 cases should have priority?

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+                console.log('in catch for getLock')
+                console.error(err)
+            })
+    }
+
+    async function toggleMetaFailureBanner(shouldShow) {
+        const bannerId = "notification_metaFailure"
+        const bannerText = "Failed to check notebook status, changes may not be saved to workspace. Retrying..."
+
+        if ($("#" + bannerId).length > 0) {
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: replace with removeModeBanner()

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+                    const message = getLockConflictBody(body.lockedBy)
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => `<p>This file is currently being editted by ${lockHolder}.</p><br/><p>You can make a copy, or run it in Playground Mode to explore and execute its contents without saving any changes.`
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: Split into multiple lines

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
```
Comment (jmthibault79) [inline docker/jupyter/custom/JupyterLockingExtension.js]: Why do we need to remove `notebooks/` here?

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 8 above)
Comment (jmthibault79) [reply docker/jupyter/custom/JupyterLockingExtension.js]: I mean the debugging lines like this one.  Keep the informative ones.

File: src/main/resources/jupyter/safe-mode.js (hunk 3 above)
Comment (jdcanas) [reply src/main/resources/jupyter/safe-mode.js]: I think it may be worth polling to design for the possibility of switching in the future. 

As its currently envisioned,  when a user wishes to transfer from playground mode to edit mode or vice versa it will result in notebook reload and thus the plugin will re-trigger, so polling is not needed  as is.

I'm open to removing it for now, but I think the performance overhead would be negligible.

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 4 above)
Comment (jdcanas) [reply docker/jupyter/custom/JupyterLockingExtension.js]: Still pending discussion with Adrian on the styling of buttons outside of the lock modal. Didn't want to add styling when its possible for the UX to change. One of the other buttons may receive this class, but im not sure which one will be considered the 'primary'

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 9 above)
Comment (jdcanas) [reply docker/jupyter/custom/JupyterLockingExtension.js]: Not sure I understand, they're mutually exclusive

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 12 above)
Comment (jdcanas) [reply docker/jupyter/custom/JupyterLockingExtension.js]: The jupyter contents API requires the local path to the file, but it assumes all notebooks are within notebooks/ and wants it omitted

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 7 above)
Comment (jdcanas) [reply docker/jupyter/custom/JupyterLockingExtension.js]: the Jupyter server-side api required to save the file, which is needed to change between edit and playground mode from within the plugin

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 5 above)
Comment (jdcanas) [reply docker/jupyter/custom/JupyterLockingExtension.js]: Do you have any idea how to access this cookie or how I can test this? This has been something I haven't been sure how to test

**Comment by jdcanas** [comment]
Still coming up with a testing strategy... the stuff hitting the Jupyter server I should be able to test by spinning up the proper docker image. For an e2e test, I will most likely need a fiab with both welder fully-functional and the jupyter image with this plugin. 

I'm hoping to merge this in this 'draft' state to allow future PRs to move forward with docker image/welder integration.

Currently, I tested all of this with a local mock server that returns whatever payloads/status codes I needed to test all use/edge cases.

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 13]
```diff
@@ -0,0 +1,327 @@
+const dialog = require('base/js/dialog')
+const utils = require("base/js/utils")
+
+define(() => {
+    var modalOpen = false
```
Comment (calbach) [inline docker/jupyter/custom/JupyterLockingExtension.js]: let?

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+            checkMeta()
+        }, lastLockedTimer)
+
+        window.onbeforeunload(function() {
+            clearInterval(syncMaintainer)
+        })
+    }
+
+    async function checkMeta() {
+        const payload = Object.assign(basePayload, {
```
Comment (calbach) [inline docker/jupyter/custom/JupyterLockingExtension.js]: This mutates basePayload - I don't think that's what you want. splay shallow copy is probably good here:


```
fetch(url, {
  ...basePayload,
  body: JSON.stringify(..
}
```

File: docker/jupyter/custom/JupyterLockingExtension.js [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+            body: JSON.stringify(getLocalPath()),
+            method: 'POST'
+        })
+
+        return fetch(checkMetaUrl, payload)
+            .then(res => {
+                console.log('in first then of checkmeta', res)
+                if (!res.ok && res.status == 412) {
+                    promptUserWithModal(syncIssueTitle, noRemoteFileButtons, syncIssueFatalBody)
```
Comment (calbach) [inline docker/jupyter/custom/JupyterLockingExtension.js]: 412 here means no storage link. This is expected for a file which is not controlled by a storage link. THe extension should bail out at this point and not attempt to check metadata, lock, or do anything else.

File: src/main/resources/jupyter/safe-mode.js (hunk 3 above)
Comment (jmthibault79) [reply src/main/resources/jupyter/safe-mode.js]: ok to keep

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 4 above)
Comment (jmthibault79) [reply docker/jupyter/custom/JupyterLockingExtension.js]: ok

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 9 above)
Comment (jmthibault79) [reply docker/jupyter/custom/JupyterLockingExtension.js]: ha.  that is true.  never mind.

File: docker/jupyter/custom/JupyterLockingExtension.js (hunk 7 above)
Comment (jmthibault79) [reply docker/jupyter/custom/JupyterLockingExtension.js]: OK.  this will also have to be proxied thru Leo, since the UI won't have access to anything server-side.

**Comment by jmthibault79** [comment]
Please rebase onto develop - and generally, do that instead of merging.  Merging adds extra commits to the PR that make it more confusing.

File: src/main/resources/jupyter/safe-mode.js (hunk 3 above)
Comment (calbach) [reply src/main/resources/jupyter/safe-mode.js]: In Jupyter, it should not be possible to change notebooks within a single client-side session. Technically a notebook could change by way of /storageLinks, but this should never happen and would probably indicate deeper bugs if it did.

I would just lock in whatever mode you get back from the first checkMetadata request.

File: src/main/resources/jupyter/edit-mode.js [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+    const modeBannerId = "notification_mode"
+    const lockConflictTitle = "File is in use"
+    const syncIssueTitle = "File versions out of sync"
+    const syncIssueBody = "Your version of this file does not match the version in the workspace. What would you like to do?"
+    const syncIssueFatalBody = "This file was either deleted or never was stored with us."
+
+    //TODO URL resolution
+    const leoUrl = ''
+        // const leoUrl = 'http://localhost:8081'
+    const welderUrl = leoUrl + `/proxy/${googleProject}/${clusterName}/welder`
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: Probably easier and can avoid the templating complexity: grab `Jupyter.notebook.base_url`, then replace "/notebooks/" with "/welder/"

File: src/main/resources/jupyter/edit-mode.js [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+                if (!res.ok && res.status == 412) {
+                    promptUserWithModal(syncIssueTitle, noRemoteFileButtons, syncIssueFatalBody)
+                } else if (!res.ok && res.status >= 400) {
+                    throw Error(res.statusText)
+                }
+                return res.json()
+            })
+            .then(res => {
+                console.log('res in check meta after .json()', res)
+                handleCheckMetaResp(res) //sets meta state
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: optional: if lockedBy != null we could show the locked dialog at this point, rather than trying to acquire it

File: src/main/resources/jupyter/edit-mode.js [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+            })
+            .catch(err => {
+                console.log('in checkMeta catch')
+                console.error(err)
+                removeElementById(modeBannerId)
+                toggleMetaFailureBanner(true)
+            })
+    }
+
+    //this function assumes any status not included in these lists represents an in sync notebook to defend against future fields being added being auto-categorized as failures
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: It would be better to autocategorize future statuses as failures. LIVE should be the only good status for the foreseeable future. This is a good thing to call out though - luckily Welder and the UI extension are deployed in tandem so we should be pretty insulated against server/client skew in this case.

File: src/main/resources/jupyter/edit-mode.js [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+    function handleCheckMetaResp(res) {
+        meta = res //set meta state
+
+        const outOfSyncStatuses = ["DESYNCHRONIZED", "LOCAL_CHANGED", "REMOTE_CHANGED"]
+        const fatalSyncStatuses = ["REMOTE_NOT_FOUND"]
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: THis is actually less fatal than the other cases - I'd probably rename. The user should be able to continue and just save normally.

File: src/main/resources/jupyter/edit-mode.js [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => {
+        return `<p>This file is currently being editted by ${lockHolder}.</p>` +
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: edited

Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: Unfortunately we won't have the specifics here (per the user metadata privacy doc) - will just have to say "by another user"

File: src/main/resources/jupyter/edit-mode.js [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
+
+        console.log('switching to playground path: ', newPath)
+
+        const payload = {
+            ...basePayload,
+            method: 'PATCH',
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: What are we patching here? Can you add a comment? The semantics I'd expect here are:

- copy or "save as" to playground mode directory
- redirect to new file

File: src/main/resources/jupyter/edit-mode.js [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: "notebooks" is not guaranteed to be in this string; to form the playground mode path:

- take full relative notebook path, e.g. dir/subdir/a.ipnyb
- replace localBaseDirectory with localSafeModeBaseDirectory (should both be prefixes of the path)

File: src/main/resources/jupyter/edit-mode.js [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+        const payload = {
+            ...basePayload,
+            method: 'PATCH',
+            body: JSON.stringify({ path: newPath })
+        }
+
+        console.log('url: ', url, newPath)
+
+        fetch(url, payload).then(res => {
+            window.location.href = jupyterBaseUrl + "notebooks/" + newPath
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: You should be able to simplify all of this just using relative paths, e.g. `window.location.href = "/notebooks/" + newPath`

If there's a basePath set, you could even go potentially simpler: `window.location.href = newPath`

File: src/main/resources/jupyter/init-actions.sh [hunk 24]
```diff
@@ -323,6 +324,15 @@ if [[ "${ROLE}" == 'Master' ]]; then
         docker cp /etc/${JUPYTER_SAFE_MODE_PLUGIN} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
       fi
 
+      # If edit-mode.js was specified, copy it into the jupyter docker container.
+        if [ ! -z ${JUPYTER_EDIT_MODE_PLUGIN_URI} ] ; then
+          log 'Installing edit-mode.js extension...'
+          gsutil cp ${JUPYTER_EDIT_MODE_PLUGIN_URI} /etc
+          JUPYTER_EDIT_MODE_PLUGIN=`basename ${JUPYTER_EDIT_MODE_PLUGIN_URI}`
+          retry 3 docker exec ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
```
Comment (rtitle) [inline src/main/resources/jupyter/init-actions.sh]: Just curious, is there a reason to use `~/.jupyter/custom` instead of installing this as a "normal" extension (see the block starting on L244)?

Also I'm wondering if this works out of the box alongside `google_sign_in.js` (which is also in `~/jupyter/custom`). Am wondering if this needs updating to reference `edit-mode.js` and `safe-mode.js`: https://github.com/DataBiosphere/leonardo/blob/develop/src/main/resources/jupyter/google_plugin_jupyter.js

File: src/main/resources/jupyter/edit-mode.js (hunk 16 above)
Comment (jmthibault79) [reply src/main/resources/jupyter/edit-mode.js]: That doesn't match the Leonardo/Welder proxy URL pattern (see [cluster-site.conf](https://github.com/DataBiosphere/leonardo/blob/develop/src/main/resources/jupyter/cluster-site.conf).

Jupyter -> `leoBaseUrl/notebooks/${googleProject}/${clusterName}/jupyter`
Welder ->  `leoBaseUrl/proxy/${googleProject}/${clusterName}/welder/${pathToWelderEndpoint}`

We could make it more symmetrical as you suggest, but this would require adding a new proxy route to Leonardo, in addition to updating cluster-site.conf.  I chose the `/proxy/` pattern simply because that was easier: cluster-site.conf only.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 25]
```diff
@@ -120,24 +120,13 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
 
-  // Startup scripts to install on the cluster master node.
-  // This allows Jupyter and Welder to start back up after a cluster is resumed.
+  // Startup script to install on the cluster master node. This allows Jupyter to start back up after
+  // a cluster is resumed.
+  //
+  // The || clause is included because older clusters may not have the run-jupyter.sh script installed,
+  // so we need to fall back running `jupyter notebook` directly. See https://github.com/DataBiosphere/leonardo/issues/481.
   private lazy val masterInstanceStartupScript: immutable.Map[String, String] = {
-    val googleKey = "startup-script"  // required; see https://cloud.google.com/compute/docs/startupscript
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Seems like you removed the welder startup script. Will that be re-added in a different PR?

File: src/main/resources/jupyter/edit-mode.js (hunk 16 above)
Comment (calbach) [reply src/main/resources/jupyter/edit-mode.js]: OK, this seems fine then.

File: src/main/resources/jupyter/edit-mode.js (hunk 17 above)
Comment (jdcanas) [reply src/main/resources/jupyter/edit-mode.js]: There was a discussion around this, and we determined that it would be best to leave it up to the API to determine the lock status, as its best not to trust client-side logic to determine something that could impact other users. Additionally, it allows looser coupling.

File: src/main/resources/jupyter/edit-mode.js [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+                } else if (!res.ok && res.status >= 400) {
+                    throw Error(res.statusText)
+                }
+                return res.json()
+            })
+            .then(res => {
+                console.log('res in check meta after .json()', res)
+                handleCheckMetaResp(res) //sets meta state
+                toggleMetaFailureBanner(false) //sets banner for meta status
+                maintainLock(res) //gets lock
```
[the remark this answers is not in the document]
Comment (jdcanas) [reply src/main/resources/jupyter/edit-mode.js]: It does only get the lock if res.syncMode == 'Edit', but it needs to display a mode banner regardless, as we are collapsing that portion of the safe-mode extension into this extension to avoid the possibility of conflicting banners. I renamed the method to maintainLockState to clarify.

File: src/main/resources/jupyter/edit-mode.js (hunk 17 above)
Comment (calbach) [reply src/main/resources/jupyter/edit-mode.js]: I'm not suggesting any client-side logic. The Welder API has a `lastLockedBy` field returned in the metadata response, it was supposed to be null if you can lock - non-null if locked by someone else. If that's changed, we should stop returning it in the response JSON and update the design doc.

File: src/main/resources/jupyter/edit-mode.js (hunk 26 above)
[the remark this answers is not in the document]
Comment (calbach) [reply src/main/resources/jupyter/edit-mode.js]: OK - in my mind safe mode and locking are unrelated, so it doesn't make sense to me that `maintainLock` results in rendering of the banner (for edit or safe mode). I might suggest restructuring the tree of function calls here, but agreed it looks like the current end behavior is correct.

File: src/main/resources/jupyter/edit-mode.js (hunk 22 above)
Comment (jdcanas) [reply src/main/resources/jupyter/edit-mode.js]: yeah, this logic was messy because the Jupyter server automatically prefixes the path to all notebooks with notebooks/ with my local config. I expected once I get it running in the docker image that this will change.

File: src/main/resources/jupyter/edit-mode.js (hunk 23 above)
Comment (jdcanas) [reply src/main/resources/jupyter/edit-mode.js]: As I understand it, we will need to call the leo jupyter proxy endpoint (jupyterBaseUrl), which will then proxy the call to the Jupyter server at the API path specified after the leo path. I had this working locally on my Jupyter server, but this will need more testing once I have the extension in the jupyter image. 

That being said, I didn't know you could specify a relative path when assigning window.location.href, so I'll investigate simplifying this

File: src/main/resources/jupyter/edit-mode.js (hunk 21 above)
Comment (jdcanas) [reply src/main/resources/jupyter/edit-mode.js]: added

File: src/main/resources/jupyter/edit-mode.js [hunk 27]
```diff
@@ -0,0 +1,404 @@
...
+
+    const modeBannerId = "notification_mode"
+    const lockConflictTitle = "File is in use"
+    const syncIssueTitle = "File versions out of sync"
+    const syncIssueBody = "Your version of this file does not match the version in the workspace. What would you like to do?"
+    const syncIssueNotFoundBody = "This file was either deleted or never was stored with us."
+
+    //TODO URL resolution
+    // const leoUrl = ''
+    const leoUrl = 'http://localhost:8080'
```
Comment (rtitle) [inline src/main/resources/jupyter/edit-mode.js]: This is for development purposes I assume.

File: src/main/resources/reference.conf [hunk 28]
```diff
@@ -25,10 +25,11 @@ clusterResources {
   proxyDockerCompose = "proxy-docker-compose.yaml"
   welderDockerCompose = "welder-docker-compose.yaml"
   proxySiteConf = "cluster-site.conf"
-  googleSignInJs = "google_sign_in.js"
-  jupyterGooglePlugin = "google_plugin_jupyter.js"
+  googleSignInJs = "google_sign_in.js" //TODO make naming convention for nbextension js files
+  jupyterGooglePlugin = "google_plugin_jupyter.js" //TODO rename this because its name is misleading as to its actual purpose?
```
Comment (rtitle) [inline src/main/resources/reference.conf]: +1 on renaming this, it's really an entrypoint for ALL js extensions, not just the google one

File: src/main/resources/reference.conf [hunk 29]
```diff
@@ -25,10 +25,11 @@ clusterResources {
   proxyDockerCompose = "proxy-docker-compose.yaml"
   welderDockerCompose = "welder-docker-compose.yaml"
   proxySiteConf = "cluster-site.conf"
-  googleSignInJs = "google_sign_in.js"
-  jupyterGooglePlugin = "google_plugin_jupyter.js"
+  googleSignInJs = "google_sign_in.js" //TODO make naming convention for nbextension js files
```
Comment (rtitle) [inline src/main/resources/reference.conf]: +1 on refactoring this config. Instead of individual configs for each extension, could have something like:
```
# the entrypoint
jupyterPlugin = "jupyter_plugin.js"
# the actual list of extensions which get templated into jupyter_plugin.js
jupyterExtensions = [ "list", "of", "jupyter", "extensions" ]
```
I haven't thought this through deeply and could probably wait for a separate refactoring ticket outside of this PR.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 30]
```diff
@@ -952,14 +956,19 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
 
     for {
       // Upload the init script to the bucket
       _ <- leoGoogleStorageDAO.storeObject(initBucketName, GcsObjectName(clusterResourcesConfig.initActionsScript.value), initScriptContent, "text/plain")
 
-      // Upload the googleSignInJs file to the bucket
+      // Upload the nb extensions to the bucket
       _ <- leoGoogleStorageDAO.storeObject(initBucketName, GcsObjectName(clusterResourcesConfig.googleSignInJs.value), googleSignInJsContent, "text/plain")
+      _ <- leoGoogleStorageDAO.storeObject(initBucketName, GcsObjectName(clusterResourcesConfig.jupyterEditModePlugin.value), editModeJsContent, "text/plain")
+      _ <- leoGoogleStorageDAO.storeObject(initBucketName, GcsObjectName(clusterResourcesConfig.jupyterSafeModePlugin.value), safeModeJsContent, "text/plain")
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: It _seems_ like the templating should be occurring based on this code. Are you not seeing it working?

File: src/main/resources/reference.conf (hunk 28 above)
Comment (jdcanas) [reply src/main/resources/reference.conf]: done

File: src/main/resources/reference.conf (hunk 29 above)
Comment (jdcanas) [reply src/main/resources/reference.conf]: +1 on doing this in another PR

File: src/main/resources/jupyter/edit-mode.js (hunk 21 above)
Comment (calbach) [reply src/main/resources/jupyter/edit-mode.js]: Per offline discussion: want copy rather than move, here and below (save as)

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 31]
```diff
@@ -120,8 +120,11 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     network = dataprocConfig.vpcNetwork.map(VPCNetworkName),
     targetTags = List(NetworkTag(dataprocConfig.networkTag)))
 
-  // Startup scripts to install on the cluster master node.
-  // This allows Jupyter and Welder to start back up after a cluster is resumed.
+  // Startup script to install on the cluster master node. This allows Jupyter to start back up after
+  // a cluster is resumed.
+  //
+  // The || clause is included because older clusters may not have the run-jupyter.sh script installed,
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: I notice the comment `The || clause ...` is duplicated down below. Probably can remove it here.

File: src/main/resources/jupyter/edit-mode.js [hunk 32]
```diff
@@ -0,0 +1,396 @@
+const dialog = require('base/js/dialog')
+const utils = require("base/js/utils")
+
+define(() => {
+    // TEMPLATED CODE
+    // Leonardo has logic to find/replace templated values in the format $(...).
+    const googleProject = $(googleProject);
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: I don't think Leo will subtitute quotes for you, should probably be `"$(googleProject)"`

File: src/main/resources/jupyter/edit-mode.js [hunk 33]
```diff
@@ -0,0 +1,396 @@
...
+        mode: "no-cors",
+        headers: headers
+    }
+
+    const jupyterBaseUrl = `/notebooks/${googleProject}/${clusterName}`
+    const jupyterContentsAPIUrl = jupyterBaseUrl + "/api/contents/"
+
+    function init() {
+        console.log('here in edit mode init')
+            // checkMeta()
```
Comment (calbach) [inline src/main/resources/jupyter/edit-mode.js]: Be sure to revert this before merging - extension doesn't do anything without this
